nixos/modules/installer/kexec/kexec-boot.nix: move into nixos/lib, expose `kexec` nixos/release.nix

flokli opened the pull request:
`nixos/modules/installer/kexec/kexec-boot.nix` doesn't contain any
custom NixOS config, other than importing `netboot-minimal.nix`.

It's a bit annoying to use - you need to import that file into your
system configuration, to then get an attribute pointing to the netboot
artifacts.

Move the function into nixos/lib/make-kexec.nix, so it can be used
outside NixOS configuration context.

This allows using it from nixos/release.nix, and expose a
`kexec.$system`, like we already have `netboot.$system`.

Once that's all in place, we document its usage, which now is pleasingly
similar to creating iPXE artifacts.

Continuation of https://github.com/NixOS/nixpkgs/pull/161253#issuecomment-1146865062.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

roberth commented on nixos/doc/manual/installation/installing-kexec.section.md:
````diff
@@ -0,0 +1,59 @@
+Instead of using the default installer image, you can also pass in a custom
+configuration:
+
+```nix
+
+{ system ? builtins.currentSystem }:
+
+{
+  myNetboot = makeKexec {
+    module = ./path/to/my/config.nix;
+    inherit system;
+  };
+}
+```
````
+1 for adding to `release.nix`, so we have ready-built default images, but why not keep the module? 

`makeKexec` is a let binding, which is only accessible by Nixpkgs forks.

You could use the module in `release.nix`. That will keep `release.nix` simple and serve as a reasonable example of how to build one in an expression, for people who prefer to dive into code and learn by example.

I guess the alternative is to expose `makeKexec`, but that would create yet another entrypoint into NixOS, which is hostile to composition.

This seems to be a working example with the module:

```nix
let
  kexecConfig = import <nixpkgs/nixos> {
    configuration = { modulesPath, ... }: {
      imports = [
        (modulesPath + "/installer/kexec/kexec-boot.nix")
        (modulesPath + "/installer/netboot/netboot-minimal.nix")
        # for example
        ./configuration.nix
      ];
    };
  };
in
kexecConfig.config.system.build.kexecBoot
```

or perhaps a command-line alternative

`configuration.nix`
```
{ modulesPath, ... }: {
  imports = [
    (modulesPath + "/installer/kexec/kexec-boot.nix")
    (modulesPath + "/installer/netboot/netboot-minimal.nix")
  ];
};
```

```
nix-build '<nixpkgs/nixos>' \
  --arg configuration ./configuration.nix \
  --attr config.system.build.kexecBoot
```

roberth replied:
It is still possible to provide a simplified interface, as previously discussed in https://github.com/NixOS/nixpkgs/pull/161253#discussion_r811306610

I've been somewhat critical of it because of eval performance, but the simplicity of such an interface is very much worth it for use cases where the kexec variant isn't evaluated often. https://github.com/NixOS/rfcs/pull/22 would make it a no-brainer.

flokli commented:
I marked this as a draft, because I realize just having it in `nixos/release.nix` makes it very hard to be used from elsewhere.

I'll push a new version of this in a bit.

flokli commented:
I pushed a new version, which now also fixes the tests. The commit message and PR description is also updated.

I'm still not entirely sure if I like this more.

My main annoyance with the current approach in `master` was that the function wasn't really exposed standalone, it wasn't really possible to use this without importing the profile, and just having that "trampoline profile" felt a bit silly too.

I'm still a bit unsure if I like the current state. Maybe only having it availabe in the profile is fine, as we need the profile due to all the underlying system config requirements anyways?

Another idea would be to just move the kexecTree generation into `nixos/modules/installer/netboot/netboot.nix`, alongside the other functions there - and have `nixos/release.nix` use them, instead of writing its own `makeNetboot`?

roberth commented on nixos/doc/manual/installation/installing-kexec.section.md:
```diff
@@ -0,0 +1,69 @@
+# "Booting" into NixOS via kexec (PXE) {#sec-booting-via-kexec}
```
> (PXE)

Maybe I missed something, but there's no PXE involved here, is there?

flokli replied:
yes, you're right.

roberth commented:
I'm not sure what to think about the refactor. I don't like what's in `nixos/lib`, a combination of public interfaces and "utils" sprawl, so it's hard for me to like the change.

A more real problem is that the use case still depends on `netboot` stuff, which it shouldn't. It's also weird to use stuff from the `modules/installer` directory for things that aren't installer images. We should extract some of those modules' contents into new modules that the installer modules import.

The scope for this PR is mostly just adding documentation, so it already qualifies as good enough.

So I'm kinda ok with it. Perhaps someone who uses this can do the final review?

flokli commented:
`netboot` is a bit too specific. Both PXE and kexec just build a kernel and initrd that entirely run from inside initrd.

`nixos/modules/installer/netboot/netboot.nix` even copied some of the comments when copying the squashfs/overlayfs bits from `nixos/modules/installer/cd-dvd/iso-image.nix` - see the "to make this a live CD" comment :laughing: 

We might rename `netboot*` with something else, or factor out all these bits into something shared, i don't know. I also don't know about a better name.

I'm not happy with the refactor either. I'll try the alternative approach from https://github.com/NixOS/nixpkgs/pull/177010#issuecomment-1151106562, undoing the move of the code to `nixos/lib`, keeping a `system.build.kexecBoot` attribute, but defining it in`nixos/modules/installer/netboot/netboot.nix`

I'll leave renaming `netboot*.nix` to a later PR.

flokli committed:
nixos/…/kexec-boot.nix: move into netboot.nix, rename to kexecTree

`nixos/modules/installer/kexec/kexec-boot.nix` doesn't contain any
custom NixOS config, other than importing `netboot-minimal.nix` (which
imports `netboot-base.nix`, which imports `netboot.nix`.

`netboot.nix` really is just describing a self-contained system config,
running entirely off kernel and initrd, so we might as well move the
kexec script generation there as well.

`netboot.nix` already contains some `system.build` attributes.
Provide a `system.build.kexecTree` attribute (and `kexecScript` for
composability).

flokli committed:
nixos/release.nix: expose a `kexec.$system` attribute

flokli committed:
nixos/doc: document how to use kexecTree

This documents how to build kexec artifacts.

flokli commented:
@roberth I pushed the other version of this.

The function is now back in `nixos/modules`, and there's `system.build.kexecTree` and `system.build.kexecScript`attributes.

Indeed this makes things much easier accessible, so I updated the docs accordingly. PTAL.

roberth commented:
Looks ok.

> rename

We could rename `system.build.netbootRamdisk` to `system.build.wholeOSRamdisk` and extract its dependencies into a similarly named module.
